### PR TITLE
[go] Bump images, dependencies and versions to go 1.21.0

### PR DESCRIPTION
**Pull request by cpanato**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

- Bump images, dependencies and versions to go 1.21.0

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/release/issues/3076

#### Does this PR introduce a user-facing change?

```release-note
Kubernetes is now built with go 1.21.0
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by cpanato** [comment]
cc @liggitt @ameukam

**Comment by MadhavJivrajani** [comment]
cc @akhilerm

**Comment by MadhavJivrajani** [comment]
From `pull-kubernetes-verify`
```
go: found k8s.io/code-generator in k8s.io/code-generator v0.0.0-00010101000000-000000000000
../code-generator/tools.go:25:2: import "k8s.io/code-generator/cmd/applyconfiguration-gen" is a program, not an importable package
```
A smaller reproducer of this is to run the following in root of `k/k`:
```
go list -tags=tools all
```
On running this, I get:
```
❯ go list -tags=tools all
build/tools.go:25:2: import "github.com/onsi/ginkgo/v2/ginkgo" is a program, not an importable package
```
Seems like this is tracked here: https://github.com/golang/go/issues/59186
Working on a workaround.

**Comment by MadhavJivrajani** [comment]
Tracking issue and WIP PR for go1.21 support for `golangci-lint`:
- https://github.com/golangci/golangci-lint/issues/3933
- https://github.com/golangci/golangci-lint/pull/3922

Thread on gophers slack with the golangci-lint maintainer: https://gophers.slack.com/archives/CS0TBRKPC/p1688365821502669?thread_ts=1686144370.506279&cid=CS0TBRKPC

**Comment by cpanato** [comment]
@saschagrunert @liggitt looks like using bookworm is not working all properly, but not 100% sure

**Comment by saschagrunert** [comment]
It looks like that `go list -tags=tools all` does not work any more with 1.21:

```
=== computing tools imports for k8s.io/apiextensions-apiserver
+ go list -tags=tools all
go: found k8s.io/code-generator in k8s.io/code-generator v0.0.0-00010101000000-000000000000
../code-generator/tools.go:25:2: import "k8s.io/code-generator/cmd/applyconfiguration-gen" is a program, not an importable package
```

It also does not work with rc2, independently if we use bookworm or bullseye images.

Looks like that we cannot import main binaries any more in golang.

We could switch to package imports as alternative, for example:

```diff
-       _ "github.com/onsi/ginkgo/v2/ginkgo"
+       _ "github.com/onsi/ginkgo/v2"
```

But this has other implications. Would be interesting to hear the thoughts from @kubernetes/sig-architecture here.

**Comment by MadhavJivrajani** [comment]
@saschagrunert yeah :/
See https://github.com/kubernetes/kubernetes/pull/118996#issuecomment-1617386511
The "recommended" way is using `go list -e` (https://github.com/golang/go/issues/59186#issuecomment-1584499117)

Trying that out here: https://github.com/kubernetes/kubernetes/pull/119027

**Comment by cpanato** [comment]
As a follow up I will start the work to build the images for k8s 1.29, but as this is ready now lets move forward with this and later I will update

File: build/build-image/cross/VERSION
Comment (liggitt) [inline build/build-image/cross/VERSION:1]: interesting, I would have expected this to already be at 1.28.0-go1.20.7...

File: build/build-image/cross/VERSION
Comment (liggitt) [inline build/build-image/cross/VERSION:1]: shouldn't this be 1.29.0-go1.21.0-...?

File: build/dependencies.yaml
Comment (liggitt) [inline build/dependencies.yaml:140]:
```suggestion
    version: v1.29.0-go1.21.0-bookworm.0
```

File: build/build-image/cross/VERSION
Comment (cpanato) [reply build/build-image/cross/VERSION:1]: fixed

File: build/common.sh
Comment (saschagrunert) [inline build/common.sh:100]:
```suggestion
readonly __default_go_runner_version=v2.3.1-go1.21.0-bookworm.0
```

File: build/common.sh
Comment (MadhavJivrajani) [inline build/common.sh:100]: I see this failure in `pull-kubernetes-dependencies`:
```
INFO Validating zeitgeist options...              
ERRO /home/prow/go/src/k8s.io/kubernetes/build/dependencies.yaml indicates that registry.k8s.io/go-runner: dependents should be at version v2.3.1-go1.21.0-bullseye.0, but the following files didn't match: build/common.sh 
FATA Dependencies are not in sync   
```
The `go_runner` image still uses go1.21.0-bullseye iiuc: 
https://github.com/kubernetes/release/pull/3212/files#diff-5f5e895c5bc907a3e7049a67d1513bc984dff0a31cbff0b69b9aac7c6114c93a

Comment (cpanato) [reply build/common.sh:100]: thanks!

File: build/dependencies.yaml
Comment (liggitt) [inline build/dependencies.yaml:137]: isn't this ready to uncomment now?

Comment (cpanato) [reply build/dependencies.yaml]: forgot that, thanks

**Comment by liggitt** [comment]
/lgtm
/approve

if CI is happier this time :)

**Comment by pacoxu** [comment]
/hold 
the CI failure seems to be valid.
```
Aug 31 16:49:01 kind-control-plane systemd[1]: Started kubelet: The Kubernetes Node Agent.
Aug 31 16:49:01 kind-control-plane kubelet[522]: /usr/bin/kubelet: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by /usr/bin/kubelet)
Aug 31 16:49:01 kind-control-plane kubelet[522]: /usr/bin/kubelet: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by /usr/bin/kubelet)
Aug 31 16:49:01 kind-control-plane systemd[1]: kubelet.service: Main process exited, code=exited, status=1/FAILURE
Aug 31 16:49:01 kind-control-plane systemd[1]: kubelet.service: Failed with result 'exit-code'.
```
kubelet failed to start. https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/118996/pull-kubernetes-e2e-kind-ipv6/1697285760991891456/artifacts/kind-control-plane/kubelet.log

**Comment by liggitt** [comment]
Did the build images bump to an OS version with too new of a glibc?

**Comment by MadhavJivrajani** [comment]
Potentially this: https://github.com/kubernetes/release/pull/3212 ?

cc @saschagrunert - fyi ^

**Comment by MadhavJivrajani** [comment]
I see go-runner is still on bullseye, any reason why that's not on bookworm too?

https://github.com/kubernetes/release/pull/3212/files#diff-5f5e895c5bc907a3e7049a67d1513bc984dff0a31cbff0b69b9aac7c6114c93a

**Comment by cpanato** [comment]
> I see go-runner is still on bullseye, any reason why that's not on bookworm too?
> 
> https://github.com/kubernetes/release/pull/3212/files#diff-5f5e895c5bc907a3e7049a67d1513bc984dff0a31cbff0b69b9aac7c6114c93a

we are using bookworm
https://github.com/kubernetes/release/blob/master/images/build/go-runner/variants.yaml#L2-L9

and here in https://github.com/kubernetes/kubernetes/pull/118996/files#diff-64833c91e011032ba2c7a5d17a52dd97ff3d10ecc644ffbfad15259718c75813R100

**Comment by cpanato** [comment]
> Did the build images bump to an OS version with too new of a glibc?

Looks like bookworm has some new things.

should we have both images with bullseye and bookworm?
cc @liggitt @saschagrunert

**Comment by saschagrunert** [comment]
Bookworm uses a newer version of glibc compared to bullseye. I assume that the issue is that we still use bullseye in `krte` for https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/118996/pull-kubernetes-e2e-kind-ipv6/1697285760991891456/artifacts/kind-control-plane/kubelet.log

https://github.com/kubernetes/test-infra/blob/7f9d95cdc7c863d4c6fc230bb68bf82d77992e8a/images/krte/Dockerfile#L19

I'm proposing the krte bump in https://github.com/kubernetes/test-infra/pull/30603 for discussion.

**Comment by saschagrunert** [comment]
For `pull-kubernetes-e2e-gce`, it looks like that we're using `projects/ubuntu-os-cloud/global/images/ubuntu-2204-jammy-v20220712a`, which may require an update.

**Comment by cpanato** [comment]
move back to bullseye the kube-cross

cc @liggitt @saschagrunert

**Review by saschagrunert — APPROVED** [review]

**Comment by cpanato** [comment]
i saw that @pacoxu unhold so will lift

/unhold

**Comment by BenTheElder** [comment]
xref: https://github.com/kubernetes/kubernetes/pull/120279

**Pull request merged**
